NIFI-4004: Use RecordReaderFactory without FlowFile.

ijokarumawak opened the pull request:
- Added RecordReaderFactory.createReader without FlowFile.
- Added validation mechanism to reject RecordReaderFactory which requires FlowFile, from a processor which does not have incoming FlowFile.
- Removed creation of temporal FlowFile to resolve Record Schema from ConsumerLease.
- Renamed variable 'alloableValue' to 'strategy' to represent its meaning better.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented:
![image](https://cloud.githubusercontent.com/assets/1107620/26689360/5db71c78-4730-11e7-866f-4ec39773b2c0.png)

The newly added validation (above image) can detect a Record Reader which can not be used with ConsumeKafkaRecord. Actually, "HWX Schema Reference Attributes" is the only one can not be used as of today, as it requires 3 attributes to identify schema.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/schema/access/SchemaAccessStrategy.java:
```diff
@@ -38,4 +40,11 @@
+    /**
+     * @return Whether this factory needs an incoming FlowFile to resolve Record Schema.
+     */
+    default boolean isFlowFileRequired() {
```
@ijokarumawak I very much agree with the idea behind this JIRA. However, I think I would approach the solution in a somewhat different manner. Rather than having a boolean indicate whether or not a FlowFile is required, I would recommend we get rid of the FlowFile being passed in to get the schema and instead just pass in a Map<String, String>. This way, if you have no FlowFile you can just pass in an empty Map. It also allows for more flexibility so that you can pass in a Map<String, String> where the schema.name may be set by the processor, for instance, instead of an empty Map.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/schema/access/SchemaAccessUtils.java:
```diff
@@ -114,46 +116,35 @@ private static boolean isSchemaRegistryRequired(final String schemaAccessValue)
-            return new HortonworksAttributeSchemaReferenceStrategy(schemaRegistry);
-        }
+    public static SchemaAccessStrategy getSchemaAccessStrategy(final String strategy, final SchemaRegistry schemaRegistry, final ProcessContext context) {
+        return getSchemaAccessStrategy(strategy, schemaRegistry, context::getProperty);
```
This is a good approach. However, on master, there is already a change that updates this to take just a PropertyContext, so I think this needs to just be rebased.

ijokarumawak commented:
@markap14 I've refactored getSchema methods from using FlowFile to using Map<String, String>. Does it look better now?

markap14 commented:
@ijokarumawak it does look good now. Unfortunately, though, it looks like it needs to be rebased. It looks like this is a rather large PR, though, that touches a lot of different points. And it looks like there are several new processors, controller services, etc. that have been merged to master since the PR was created. I wonder if it may make sense, instead of changing RecordReaderFactory to take `Map<String, String>` variables instead of `FlowFIle flowFile`, I wonder if it makes sense to add both methods to the interface, and then have a default method in the interface, such as:

```
default RecordReader createRecordReader(FlowFile flowFile, InputStream in, ComponentLog logger) throws MalformedRecordException, IOException, SchemaNotFoundException {
        return createRecordReader(flowFile == null ? Collections.emptyMap() : flowFile.getAttributes(), in, logger);
    }

    RecordReader createRecordReader(Map<String, String> variables, InputStream in, ComponentLog logger) throws MalformedRecordException, IOException, SchemaNotFoundException;

```

Thoughts?

ijokarumawak commented:
@markap14 Thanks for the suggestion, I agree with that. I've made following changes:

- Added default method at RecordReaderFactory, so that existing processors needn't change
- Reverted changes to classes those can utilize the newly added default method, PutElasticsearchHttpRecord, AbstractPutHDFSRecord, PutParquetTes, PutDatabaseRecord, FlowFileEnumerator and FlowFileTable. This reduced the volume of this PR, a little bit.
- Rebased with the latest master, and updated few new classes to meet the new RecordSetWriterFactory method signatures.

Local contrib check passed without issue. Also tested a live flow with various record readers and writers. https://gist.github.com/ijokarumawak/a6c33eef30d0cd9786eab7eeacccb7ff

I hope it's now ready to be merged, thanks!

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-service-api/src/main/java/org/apache/nifi/serialization/RecordSetWriterFactory.java:
```diff
@@ -45,20 +51,23 @@
      * @throws SchemaNotFoundException if unable to find the schema
      */
-    RecordSchema getSchema(FlowFile flowFile, RecordSchema readSchema) throws SchemaNotFoundException, IOException;
+    RecordSchema getSchema(Map<String, String> variables, InputStream content, RecordSchema readSchema) throws SchemaNotFoundException, IOException;
```
@ijokarumawak can you explain the reasoning here for providing a Map<String, String> and an InputStream? Previously, with just the FlowFile, the Writer had access only to the attributes, not the content (because it had no Process Session). I believe that is the correct abstraction. The RecordSchema from the reader already is passed in, and the FlowFile being written to will have no content, generally, so reading from the destination FlowFile wouldn't make sense. It's possible that I'm just misunderstanding the idea here, though.

markap14 commented:
@ijokarumawak thanks. Mostly looks good, but I'm concerned about the change to the writer, as I outlined in the inline comments. Can you elaborate on the idea there?

ijokarumawak commented:
@markap14 Thanks for pointing that out. I think I got mislead by the two getSchema methods in SchemaRegistryService. I accidentally removed the 2nd form during refactoring and that caused for the unnecessary InputStream sneaked into RecordSetWriterFactory.

```
    // This one had been used by readers
    public final RecordSchema getSchema(final FlowFile flowFile, final InputStream contentStream, final RecordSchema readSchema) throws SchemaNotFoundException, IOException {
    // This one had been used by writers
    public RecordSchema getSchema(final FlowFile flowFile, final RecordSchema readSchema) throws SchemaNotFoundException, IOException {
```

I pushed another commit to bring back the 2nd signature above, and change FlowFile to Map<String, String>. Compiled and contrib-check passed locally. Also tested different readers/writers with a live flow. Thanks again for your time to review!

markap14 commented:
@ijokarumawak thanks for the updated PR! All looks good at this point. +1 Have merged to master.
